doc/monitoring_containers/monitoring_containers.md

```diff
@@ -0,0 +1,271 @@
+| Config DB    | SONiC Configuration Database |
+| CLI          | Command Line Interface       |
+
+# 1 Feature Overview
+SONiC is a collection of various switch applications which are held in docker containers
+such as BGP container and SNMP container. Each application usually includes several processes which are 
+working together to provide and receive the services from other modules. As such, the health of
+critical processes in each docker container is imperative not only for the docker
+container working correctly but also for the intended functionalities of entire SONiC switch.
+
+## 1.1 Monitoring
+This feature is used to monitor the running status of critical processes and critical resource 
+usage such as CPU, memory and disk of each docker container.
+
+We used Monit system tool to detect whether a critical process is running or not and whether 
+the resource usage of a docker container is beyond the pre-defined threshold.
+
+## 1.2 Auto-Mitigating
+This feature demonstrated docker container can be automatically shut down and
+restarted if one of critical processes running in docker container exits unexpectedly. Restarting
+the entire docker container ensures that configuration is reloaded and all processes in 
+docker container get restarted, thus increasing the likelihood of entering a healthy state.
+
+We leveraged the mechanism of event listener in supervisord to auto-restart a docker container
+if one of its critical processes exited unexpectedly. We also added a configuration option to make this 
+auto-restart feature dynamically configurable. Specifically users can run CLI to configure this 
+feature residing in Config_DB as enabled/disabled status.
+
+## 1.3 Requirements
+
+### 1.3.1 Functional Requirements
+1. The Monit must provide the ability to generate an alert when a critical process has not
+    been alive for 5 minutes.
+2. The Monit must provide the ability to generate an alert when the resource usage of
+    a docker contaier is larger than the pre-defined threshold.
+3. The event listener in supervisord must receive the signal when a critical process in 
+    a docker container crashed or exited unexpectedly and then restart this docker 
+    container.
+4. CONFIG_DB can be configured to enable/disable this auto-restart feature for each docker
+    container.. 
+5. Users can access the status of auto-restart feature via the CLI utility
+    1. Users can see current auto-restart status for docker containers.
+    2. Users can configure auto-restart status for a specific docker container.
+
+### 1.3.2 Configuration and Management Requirements
+Configuration of the auto-restart feature can be done via:
+1. init_cfg.json
+2. CLI
+
+### 1.3.3 Scalability Requirements
+`Place holder`
+
+## 1.4 Design
+
+### 1.4.1 Basic Approach
+Monitoring the running status of critical processes and resource usage of docker containers
+are heavily depended on the Monit system tool. Since Monit already provided the mechanism
```

Comment:
These sentences still need touching up:

```
Monitoring the running status of critical processes and resource usage of docker containers
depends on the Monit system tool. Since Monit natively provides a mechanism
to check whether a process is running or not, it will be straightforward to integrate this into monitoring the critical processes in SONiC.
```